File: Components/DevExpressVCL/Demos/ExpressPivotGrid/Delphi/Common/DemoBasicMain.pas
unit DemoBasicMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxLookAndFeels, StdCtrls, Menus, cxCustomPivotGrid, cxExportPivotGridLink,
  cxClasses, ActnList;

type
  TfrmDemoBasicMain = class(TForm)
    mmMain: TMainMenu;
    miFile: TMenuItem;
    miExport: TMenuItem;
    miExportToExcel: TMenuItem;
    miExportToText: TMenuItem;
    miExportToHtml: TMenuItem;
    miExportToXml: TMenuItem;
    Separator1: TMenuItem;
    miExit: TMenuItem;
    miOptions: TMenuItem;
    miAbout: TMenuItem;
    SaveDialog: TSaveDialog;
    lbDescrip: TLabel;
    miTotalsPosition: TMenuItem;
    miShowTotalsForSingleValues: TMenuItem;
    miShowRowTotals: TMenuItem;
    miShowColumnTotals: TMenuItem;
    miShowColumnGrandTotals: TMenuItem;
    miShowRowGrandTotals: TMenuItem;
    miTotalsVisibility: TMenuItem;
    miShowGrandTotalsForSingleValues: TMenuItem;
    N2: TMenuItem;
    miElementsVisibility: TMenuItem;
    miShowFilterFields: TMenuItem;
    miShowColumnFields: TMenuItem;
    miShowDataFields: TMenuItem;
    miShowRowFields: TMenuItem;
    miShowFilterSeparator: TMenuItem;
    N1: TMenuItem;
    miColumnTotalsPosition: TMenuItem;
    miRowTotalsPosition: TMenuItem;
    miColumnTotalsPositionFar: TMenuItem;
    miColumnTotalsPositionNear: TMenuItem;
    miRowTotalsPositionFar: TMenuItem;
    miRowTotalsPositionNear: TMenuItem;
    miSelection: TMenuItem;
    miMultiSelect: TMenuItem;
    miHideFocusRect: TMenuItem;
    miHideSelection: TMenuItem;
    miIncludeCells: TMenuItem;
    miCrossCells: TMenuItem;
    miGrandTotalsCells: TMenuItem;
    miTotalsCells: TMenuItem;
    miRowTotalsPositionTree: TMenuItem;
    miExportToXLSX: TMenuItem;
    miTouchMode: TMenuItem;
    cxLookAndFeelController1: TcxLookAndFeelController;
    miCopyToClipboard: TMenuItem;
    ActionList1: TActionList;
    Action1: TAction;
    Action2: TAction;
    miCopyToClipboardIncludeHeaders: TMenuItem;
    miCopyToClipboardColumnHeaders: TMenuItem;
    miCopyToClipboardRowHeaders: TMenuItem;
    miCopyToClipboardRowHeadersAll: TMenuItem;
    miAllCopyToClipboardRowInnermostOnly: TMenuItem;
    miCopyToClipboardColumnHeadersAll: TMenuItem;
    InnermostOnly1: TMenuItem;
    miWYSIWYG1: TMenuItem;
    miDataOnly1: TMenuItem;
    miWYSIWYG2: TMenuItem;
    miDataOnly2: TMenuItem;
    procedure miExportToClick(Sender: TObject);
    procedure miTotalsLocationClick(Sender: TObject);
    procedure miExitClick(Sender: TObject);
    procedure miTotalsVisibilityClick(Sender: TObject);
    procedure miElementsVisibilityClick(Sender: TObject);
    procedure miAboutClick(Sender: TObject);
    procedure miMultiSelectClick(Sender: TObject);
    procedure miHideFocusRectClick(Sender: TObject);
    procedure miHideSelectionClick(Sender: TObject);
    procedure IncludeCellsClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure miTouchModeClick(Sender: TObject);
    procedure Action1Execute(Sender: TObject);
    procedure miCopyToClipboardIncludeHeadersClick(Sender: TObject);
  private
    { Private declarations }
    procedure CopyToClipboard;
  protected
    procedure AddLookAndFeelMenu; virtual;
    function GetDefaultLookAndFeelKind: TcxLookAndFeelKind; virtual;
    function GetPivotGrid: TcxCustomPivotGrid; virtual;
    function IsNativeDefaultStyle: Boolean; virtual;
    procedure SetDefaultLookAndFeel; virtual;
    procedure SyncElementsVisibilityWithMenu;
    procedure SyncTotalVisibilityWithMenu;
    procedure SyncMenuWithElementsVisibility;
    procedure SyncMenuWithOptionsSelection;
    procedure SyncMenuWithTotalsPosition;
    procedure SyncMenuWithTotalVisibility;
  public
    constructor Create(AOwner: TComponent); override;
    procedure AfterConstruction; override;

    property PivotGrid: TcxCustomPivotGrid read GetPivotGrid;
  end;

implementation

{$R *.dfm}

uses
{$IFDEF EXPRESSSKINS}
  {$I dxSkins.inc}
  dxSkinsStrs, dxSkinsForm, dxSkinInfo,
  {$IFDEF EXPRESSPAGECONTROL}
    dxSkinscxPCPainter,
  {$ENDIF}
  {$IFDEF EXPRESSSCHEDULER}
    dxSkinscxSchedulerPainter,
  {$ENDIF}
  {$IFDEF EXPRESSDOCKINGLIBRARY}
    dxSkinsdxDockControlPainter,
  {$ENDIF}
  {$IFDEF EXPRESSNAVBAR}
    dxSkinsdxNavBarPainter,
  {$ENDIF}
  {$IFDEF EXPRESSBARS}
    dxSkinsdxBarPainter,
    dxSkinsdxStatusBarPainter,
  {$IFDEF EXPRESSPAGECONTROL}
    {$IFDEF EXPRESSEDITORS}
      dxBarSkinnedCustForm,
    {$ENDIF}
  {$ENDIF}
  {$ENDIF}
{$ENDIF}
  AboutDemoForm, DemoUtils;

constructor TfrmDemoBasicMain.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
  Application.Title := Caption;
  PivotGrid.ApplyBestFit;
end;

procedure TfrmDemoBasicMain.AfterConstruction;
begin
  inherited;
  dxBarConvertMainMenu(mmMain);
  SyncMenuWithTotalsPosition;
  SyncMenuWithTotalVisibility;
  SyncMenuWithElementsVisibility;
  SyncMenuWithOptionsSelection;
end;

function TfrmDemoBasicMain.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := nil;
end;

procedure TfrmDemoBasicMain.SyncElementsVisibilityWithMenu;
begin
  with PivotGrid.OptionsView do
  begin
    ColumnFields := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowColumnFields'));
    DataFields := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowDataFields'));
    FilterFields := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowFilterFields'));
    FilterSeparator := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowFilterSeparator'));
    RowFields := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowRowFields'));
  end;
end;

procedure TfrmDemoBasicMain.SyncTotalVisibilityWithMenu;
begin
  with PivotGrid.OptionsView do
  begin
    ColumnTotals := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowColumnTotals'));
    RowTotals := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowRowTotals'));
    ColumnGrandTotals := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowColumnGrandTotals'));
    RowGrandTotals := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowRowGrandTotals'));
    TotalsForSingleValues := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowTotalsForSingleValues'));
    GrandTotalsForSingleValues := dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miShowGrandTotalsForSingleValues'));
  end;
end;

procedure TfrmDemoBasicMain.SyncMenuWithElementsVisibility;
begin
  with PivotGrid.OptionsView do
  begin
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowColumnFields'), ColumnFields);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowDataFields'), DataFields);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowFilterFields'), FilterFields);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowFilterSeparator'), FilterSeparator);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowRowFields'), RowFields);
  end;
end;

procedure TfrmDemoBasicMain.SyncMenuWithOptionsSelection;
begin
  with PivotGrid.OptionsSelection do
  begin
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miMultiSelect'), MultiSelect);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miCrossCells'), osiCrossCells in IncludeCells);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miGrandTotalsCells'), osiGrandTotalCells in IncludeCells);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miTotalsCells'), osiTotalCells in IncludeCells);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miHideFocusRect'), HideFocusRect);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miHideSelection'), HideSelection);
  end;
end;

procedure TfrmDemoBasicMain.SyncMenuWithTotalsPosition;
begin
  with PivotGrid.OptionsView do
  begin
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miColumnTotalsPositionFar'), ColumnTotalsLocation = ctlFar);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miColumnTotalsPositionNear'), ColumnTotalsLocation = ctlNear);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miRowTotalsPositionFar'), RowTotalsLocation = rtlFar);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miRowTotalsPositionNear'), RowTotalsLocation = rtlNear);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miRowTotalsPositionTree'), RowTotalsLocation = rtlTree);
  end;
end;

procedure TfrmDemoBasicMain.SyncMenuWithTotalVisibility;
begin
  with PivotGrid.OptionsView do
  begin
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowColumnTotals'), ColumnTotals);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowRowTotals'), RowTotals);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowColumnGrandTotals'), ColumnGrandTotals);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowRowGrandTotals'), RowGrandTotals);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowTotalsForSingleValues'), TotalsForSingleValues);
    dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miShowGrandTotalsForSingleValues'), GrandTotalsForSingleValues);
  end;
end;

procedure TfrmDemoBasicMain.miExportToClick(Sender: TObject);
begin
  if SaveDialog.Execute then
  begin
    case TComponent(Sender).Tag of
      2:
        cxExportPivotGridToText(SaveDialog.FileName, PivotGrid);
      3:
        cxExportPivotGridToHTML(SaveDialog.FileName, PivotGrid);
      4:
        cxExportPivotGridToXML(SaveDialog.FileName, PivotGrid);
      11:
        cxExportPivotGridDataToExcel(ChangeFileExt(SaveDialog.FileName, '.xls'), PivotGrid);
      12:
        cxExportPivotGridToExcel(SaveDialog.FileName, PivotGrid);
      51:
        cxExportPivotGridDataToExcel(ChangeFileExt(SaveDialog.FileName, '.xlsx'), PivotGrid);
      52:
        cxExportPivotGridToXLSX(SaveDialog.FileName, PivotGrid);
    end;
  end;
end;

procedure TfrmDemoBasicMain.miTotalsLocationClick(Sender: TObject);
begin
  case TComponent(Sender).Tag of
    0, 1:
      PivotGrid.OptionsView.ColumnTotalsLocation := TcxPivotGridColumnTotalsLocation(TComponent(Sender).Tag);
    2, 3, 4:
      PivotGrid.OptionsView.RowTotalsLocation := TcxPivotGridRowTotalsLocation(TComponent(Sender).Tag - 2);
  end;
  SyncMenuWithTotalsPosition;
end;

procedure TfrmDemoBasicMain.miExitClick(Sender: TObject);
begin
  Application.Terminate;
end;

procedure TfrmDemoBasicMain.miTotalsVisibilityClick(Sender: TObject);
begin
  SyncTotalVisibilityWithMenu;
end;

procedure TfrmDemoBasicMain.miTouchModeClick(Sender: TObject);
begin
  cxLookAndFeelController1.TouchMode := dxDemoIsMenuChecked(Sender);
end;

procedure TfrmDemoBasicMain.miElementsVisibilityClick(Sender: TObject);
begin
  SyncElementsVisibilityWithMenu;
end;

procedure TfrmDemoBasicMain.miAboutClick(Sender: TObject);
begin
  ShowAboutDemoForm;
end;

procedure TfrmDemoBasicMain.miMultiSelectClick(Sender: TObject);
begin
  PivotGrid.OptionsSelection.MultiSelect := not PivotGrid.OptionsSelection.MultiSelect;
end;

procedure TfrmDemoBasicMain.miHideFocusRectClick(Sender: TObject);
begin
  PivotGrid.OptionsSelection.HideFocusRect := not PivotGrid.OptionsSelection.HideFocusRect;
end;

procedure TfrmDemoBasicMain.miHideSelectionClick(Sender: TObject);
begin
  PivotGrid.OptionsSelection.HideSelection := not PivotGrid.OptionsSelection.HideSelection;
end;

procedure TfrmDemoBasicMain.IncludeCellsClick(Sender: TObject);
var
  AIncludeCells: TcxPivotGridOptionsSelectionIncludes;
begin
  dxDemoMenuItemSetChecked(Sender, dxDemoIsMenuChecked(Sender));
  AIncludeCells := [];
  if dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miCrossCells')) then
    AIncludeCells := AIncludeCells + [osiCrossCells];
  if dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miGrandTotalsCells')) then
    AIncludeCells := AIncludeCells + [osiGrandTotalCells];
  if dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miTotalsCells')) then
    AIncludeCells := AIncludeCells + [osiTotalCells];
  PivotGrid.OptionsSelection.IncludeCells := AIncludeCells;
end;

procedure TfrmDemoBasicMain.FormCreate(Sender: TObject);
begin
  SetDefaultLookAndFeel;
  AddLookAndFeelMenu;
  miTouchMode.Checked := cxIsTouchModeEnabled;
end;

procedure TfrmDemoBasicMain.AddLookAndFeelMenu;
begin
  mmMain.Items.Insert(mmMain.Items.IndexOf(miAbout), CreateLookAndFeelMenuItems(mmMain.Items
    {$IFNDEF EXPRESSBARS}, TActionList.Create(Self) {$ENDIF}));
end;

function TfrmDemoBasicMain.GetDefaultLookAndFeelKind: TcxLookAndFeelKind;
begin
  Result := lfOffice11;
end;

function TfrmDemoBasicMain.IsNativeDefaultStyle: Boolean;
begin
  Result := False;
end;

procedure TfrmDemoBasicMain.SetDefaultLookAndFeel;
begin
  dxDemoLookAndFeelController.NativeStyle := IsNativeDefaultStyle;
  dxDemoLookAndFeelController.Kind := GetDefaultLookAndFeelKind;
end;

procedure TfrmDemoBasicMain.CopyToClipboard;
begin
  GetPivotGrid.CopyToClipboard(False, miCopyToClipboardIncludeHeaders.Checked,
    miCopyToClipboardRowHeadersAll.Checked, miCopyToClipboardColumnHeadersAll.Checked);
end;

procedure TfrmDemoBasicMain.Action1Execute(Sender: TObject);
begin
  CopyToClipboard;
end;

procedure TfrmDemoBasicMain.miCopyToClipboardIncludeHeadersClick(Sender: TObject);
begin
  miCopyToClipboardColumnHeaders.Enabled := miCopyToClipboardIncludeHeaders.Checked;
  miCopyToClipboardRowHeaders.Enabled := miCopyToClipboardIncludeHeaders.Checked;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressPivotGrid/CBuilder/Common/DemoBasicMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "DemoBasicMain.h"
#include "cxExportPivotGridLink.hpp"
#include "AboutDemoForm.h"
#include "DemoUtils.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxLookAndFeels"
#pragma link "cxClasses"
#pragma resource "*.dfm"

//---------------------------------------------------------------------------
__fastcall TfrmDemoBasicMain::TfrmDemoBasicMain(TComponent* Owner)
  : TForm(Owner)
{
  Application->Title = Caption;
  FLookAndFeelController = new TcxLookAndFeelController(this);
  SyncMenuWithTotalsPosition();
  SyncMenuWithTotalVisibility();
  SyncMenuWithElementsVisibility();
  SyncMenuWithOptionsSelection();
}

TcxCustomPivotGrid* __fastcall TfrmDemoBasicMain::PivotGrid()
{
  return NULL;
}

void __fastcall TfrmDemoBasicMain::miTouchModeClick(TObject *Sender)
{
  cxLookAndFeelController1->TouchMode = miTouchMode->Checked;
}

void __fastcall TfrmDemoBasicMain::miExportToClick(TObject *Sender)
{
  if (SaveDialog->Execute()) {
	switch (((TMenuItem*)Sender)->Tag){
	  case 2:
		cxExportPivotGridToText(SaveDialog->FileName, PivotGrid(), true, "", "", "", "txt"); break;
	  case 3:
		cxExportPivotGridToHTML(SaveDialog->FileName, PivotGrid(), true, "html"); break;
	  case 4:
		cxExportPivotGridToXML(SaveDialog->FileName, PivotGrid(), true, "xml"); break;
	  case 5:
		cxExportPivotGridToXLSX(SaveDialog->FileName, PivotGrid(), true, "xlsx"); break;
	  case 11:
        cxExportPivotGridDataToExcel(ChangeFileExt(SaveDialog->FileName, ".xls"), PivotGrid(), NULL, true, true, true, false); break;
	  case 12:
		cxExportPivotGridToExcel(SaveDialog->FileName, PivotGrid(), true, true, "xls"); break;
	  case 51:
        cxExportPivotGridDataToExcel(ChangeFileExt(SaveDialog->FileName, ".xlsx"), PivotGrid(), NULL, true, true, true, false); break;
	  case 52:
		cxExportPivotGridToXLSX(SaveDialog->FileName, PivotGrid(), true, "xlsx"); break;
	}
  }
}

void __fastcall TfrmDemoBasicMain::miTotalsLocationClick(TObject *Sender)
{
	int ATag = ((TMenuItem*)Sender)->Tag;
	if (ATag < 2)
		PivotGrid()->OptionsView->ColumnTotalsLocation = TcxPivotGridColumnTotalsLocation(ATag);
	else
		PivotGrid()->OptionsView->RowTotalsLocation = TcxPivotGridRowTotalsLocation(ATag - 2);
	SyncMenuWithTotalsPosition();
}

void __fastcall TfrmDemoBasicMain::miExitClick(TObject *Sender)
{
  Application->Terminate();
}

void __fastcall TfrmDemoBasicMain::FormCreate(TObject *Sender)
{
  SetDefaultLookAndFeel();
  AddLookAndFeelMenu();
  miTouchMode->Checked = cxIsTouchModeEnabled;
}

void __fastcall TfrmDemoBasicMain::miTotalsVisibilityClick(TObject *Sender)
{
  SyncTotalVisibilityWithMenu();
}

void __fastcall TfrmDemoBasicMain::miElementsVisibilityClick(TObject *Sender)
{
  SyncElementsVisibilityWithMenu();
}
void __fastcall TfrmDemoBasicMain::miAboutClick(TObject *Sender)
{
  ShowAboutDemoForm();
}

void __fastcall TfrmDemoBasicMain::SyncElementsVisibilityWithMenu()
{
  PivotGrid()->OptionsView->ColumnFields = miShowColumnFields->Checked;
  PivotGrid()->OptionsView->DataFields = miShowDataFields->Checked;
  PivotGrid()->OptionsView->FilterFields = miShowFilterFields->Checked;
  PivotGrid()->OptionsView->FilterSeparator = miShowFilterSeparator->Checked;
  PivotGrid()->OptionsView->RowFields = miShowRowFields->Checked;
}

void __fastcall TfrmDemoBasicMain::SyncTotalVisibilityWithMenu()
{
  PivotGrid()->OptionsView->ColumnTotals = miShowColumnTotals->Checked;
  PivotGrid()->OptionsView->RowTotals = miShowRowTotals->Checked;
  PivotGrid()->OptionsView->ColumnGrandTotals = miShowColumnGrandTotals->Checked;
  PivotGrid()->OptionsView->RowGrandTotals = miShowRowGrandTotals->Checked;
  PivotGrid()->OptionsView->TotalsForSingleValues = miShowTotalsForSingleValues->Checked;
  PivotGrid()->OptionsView->GrandTotalsForSingleValues = miShowGrandTotalsForSingleValues->Checked;
}

void __fastcall TfrmDemoBasicMain::SyncMenuWithElementsVisibility()
{
  miShowColumnFields->Checked = PivotGrid()->OptionsView->ColumnFields;
  miShowDataFields->Checked = PivotGrid()->OptionsView->DataFields;
  miShowFilterFields->Checked = PivotGrid()->OptionsView->FilterFields;
  miShowFilterSeparator->Checked = PivotGrid()->OptionsView->FilterSeparator;
  miShowRowFields->Checked = PivotGrid()->OptionsView->RowFields;
}

void __fastcall TfrmDemoBasicMain::SyncMenuWithOptionsSelection()
{
  miMultiSelect->Checked = PivotGrid()->OptionsSelection->MultiSelect;
  miCrossCells->Checked = PivotGrid()->OptionsSelection->IncludeCells.Contains(osiCrossCells);
  miGrandTotalsCells->Checked = PivotGrid()->OptionsSelection->IncludeCells.Contains(osiGrandTotalCells);
  miTotalsCells->Checked = PivotGrid()->OptionsSelection->IncludeCells.Contains(osiTotalCells);
  miHideFocusRect->Checked = PivotGrid()->OptionsSelection->HideFocusRect;
  miHideSelection->Checked = PivotGrid()->OptionsSelection->HideSelection;
}

void __fastcall TfrmDemoBasicMain::SyncMenuWithTotalVisibility()
{
  miShowColumnTotals->Checked = PivotGrid()->OptionsView->ColumnTotals;
  miShowRowTotals->Checked = PivotGrid()->OptionsView->RowTotals;
  miShowColumnGrandTotals->Checked = PivotGrid()->OptionsView->ColumnGrandTotals;
  miShowRowGrandTotals->Checked = PivotGrid()->OptionsView->RowGrandTotals;
  miShowTotalsForSingleValues->Checked = PivotGrid()->OptionsView->TotalsForSingleValues;
  miShowGrandTotalsForSingleValues->Checked = PivotGrid()->OptionsView->GrandTotalsForSingleValues;
}

void __fastcall TfrmDemoBasicMain::miMultiSelectClick(TObject *Sender)
{
  PivotGrid()->OptionsSelection->MultiSelect = !PivotGrid()->OptionsSelection->MultiSelect;
  ((TMenuItem*)Sender)->Checked = PivotGrid()->OptionsSelection->MultiSelect;
}

void __fastcall TfrmDemoBasicMain::miHideFocusRectClick(TObject *Sender)
{
  PivotGrid()->OptionsSelection->HideFocusRect = !PivotGrid()->OptionsSelection->HideFocusRect;
  ((TMenuItem*)Sender)->Checked = PivotGrid()->OptionsSelection->HideFocusRect;
}

void __fastcall TfrmDemoBasicMain::miHideSelectionClick(TObject *Sender)
{
  PivotGrid()->OptionsSelection->HideSelection = !PivotGrid()->OptionsSelection->HideSelection;
  ((TMenuItem*)Sender)->Checked = PivotGrid()->OptionsSelection->HideSelection;
}

void __fastcall TfrmDemoBasicMain::IncludeCellsClick(TObject *Sender)
{
  TcxPivotGridOptionsSelectionIncludes AIncludeCells;

  AIncludeCells.Clear();
  if (miCrossCells->Checked)
	AIncludeCells << osiCrossCells;
  if (miGrandTotalsCells->Checked)
	AIncludeCells << osiGrandTotalCells;
  if (miTotalsCells->Checked)
	AIncludeCells << osiTotalCells;
  PivotGrid()->OptionsSelection->IncludeCells = AIncludeCells;
}

void __fastcall TfrmDemoBasicMain::AddLookAndFeelMenu()
{
  mmMain->Items->Insert(mmMain->Items->IndexOf(miAbout),
	CreateLookAndFeelMenuItems(mmMain->Items, FLookAndFeelController));
}

TcxLookAndFeelKind __fastcall TfrmDemoBasicMain::GetDefaultLookAndFeelKind()
{
  return(lfOffice11);
}

bool __fastcall TfrmDemoBasicMain::IsNativeDefaultStyle()
{
  return(false);
}

void __fastcall TfrmDemoBasicMain::SetDefaultLookAndFeel()
{
  FLookAndFeelController->NativeStyle = IsNativeDefaultStyle();
  FLookAndFeelController->Kind = GetDefaultLookAndFeelKind();
}
//---------------------------------------------------------------------------
void __fastcall TfrmDemoBasicMain::SyncMenuWithTotalsPosition()
{
  miColumnTotalsPositionFar->Checked = PivotGrid()->OptionsView->ColumnTotalsLocation == ctlFar;
  miColumnTotalsPositionNear->Checked = PivotGrid()->OptionsView->ColumnTotalsLocation == ctlNear;
  miRowTotalsPositionFar->Checked = PivotGrid()->OptionsView->RowTotalsLocation == rtlFar;
  miRowTotalsPositionNear->Checked = PivotGrid()->OptionsView->RowTotalsLocation == rtlNear;
  miRowTotalsPositionTree->Checked = PivotGrid()->OptionsView->RowTotalsLocation == rtlTree;
}
//---------------------------------------------------------------------------

void __fastcall TfrmDemoBasicMain::CopyToClipboard()
{
  PivotGrid()->CopyToClipboard(False, miCopyToClipboardIncludeHeaders->Checked,
	miCopyToClipboardRowHeadersAll->Checked, miCopyToClipboardColumnHeadersAll->Checked);
}
//---------------------------------------------------------------------------

void __fastcall TfrmDemoBasicMain::Action1Execute(TObject *Sender)
{
  CopyToClipboard();
}
//---------------------------------------------------------------------------

void __fastcall TfrmDemoBasicMain::miCopyToClipboardIncludeHeadersClick(TObject *Sender)
{
  miCopyToClipboardColumnHeaders->Enabled = miCopyToClipboardIncludeHeaders->Checked;
  miCopyToClipboardRowHeaders->Enabled = miCopyToClipboardIncludeHeaders->Checked;
}
//---------------------------------------------------------------------------

